velox/common/caching/SsdFile.cpp

```diff
@@ -0,0 +1,412 @@
+      int64_t score = regionScore_[i];
+      regionScore_[i] = (score * 15) / 16;
+    }
+  }
+}
+
+void SsdFile::load(SsdRun run, AsyncDataCacheEntry& entry) {
+  VELOX_CHECK_EQ(run.size(), entry.size());
+  regionScore_[regionIndex(run.offset())] += run.size();
+  ++stats_.entriesRead;
+  stats_.bytesRead += run.size();
+  if (entry.tinyData()) {
+    auto rc = pread(fd_, entry.tinyData(), run.size(), run.offset());
+    VELOX_CHECK_EQ(rc, run.size());
+  } else {
+    std::vector<struct iovec> iovecs;
+    addEntryToIovecs(entry, iovecs);
+
+    auto rc = folly::preadv(fd_, iovecs.data(), iovecs.size(), run.offset());
+    VELOX_CHECK_EQ(rc, run.size());
+  }
+  entry.setSsdFile(this, run.offset());
+}
+
+void SsdFile::read(
+    uint64_t offset,
+    const std::vector<folly::Range<char*>> buffers,
+    int32_t numEntries) {
+  stats_.entriesRead += numEntries;
+  uint64_t regionBytes = 0;
+  uint64_t regionOffset = offset;
+  // Calculate the span from first to last and the number of bytes
+  // that are used out of these. 'regionBytes' do not include
+  // gaps in the coalesced read.
+  for (auto range : buffers) {
+    if (range.data()) {
+      regionBytes += range.size();
+    }
+    regionOffset += range.size();
+  }
+  // Attribute the read bytes to the region where the read
+  // ends. Coalesced reads crossing regions are be rare and we do not
+  // consider this in tracking region read volumes.
+  regionUsed(regionIndex(offset), regionBytes);
+  stats_.bytesRead += regionBytes;
+  readFile_->preadv(offset, buffers);
+}
+
+std::pair<uint64_t, int32_t> SsdFile::getSpace(
+    const std::vector<CachePin>& pins,
+    int32_t begin) {
+  std::lock_guard<std::mutex> l(mutex_);
+  for (;;) {
+    if (writableRegions_.empty()) {
+      if (!growOrEvictLocked()) {
+        return {0, 0};
+      }
+    }
+    auto region = writableRegions_[0];
+    auto offset = regionSize_[region];
+    auto available = kRegionSize - regionSize_[region];
+    int64_t toWrite = 0;
+    for (; begin < pins.size(); ++begin) {
+      auto entry = pins[begin].entry();
+      if (entry->size() > available) {
+        break;
+      }
+      available -= entry->size();
+      toWrite += entry->size();
+    }
+    if (toWrite) {
+      // At least some pins got space from this region. If the region is is full
+      // the next call will get space from another region.
+      regionSize_[region] += toWrite;
+      return {region * kRegionSize + offset, toWrite};
+    }
+    // A region has been filled and transits from writable to
+    // evictable. Set its score to be at least the best score +
+    // kRegionSize so that it gets time to live. Otherwise it has had
+    // the least time to get hits and would be the first evicted.
+    uint64_t best = 0;
+    for (auto& score : regionScore_) {
+      best = std::max<uint64_t>(best, score);
+    }
+    regionScore_[region] = std::max(regionScore_[region], best + kRegionSize);
```

Comment:
It would be nice to decouple scoring logic. Otherwise, it is hard to reason about it and difficult to test in isolation.

Reply:
Adding a separate class for tracking would not nmake anything clearer. Can try.